File: app/src/main/java/com/boardgamegeek/filterer/PlayCountFilterer.java
package com.boardgamegeek.filterer;

import android.content.Context;
import android.support.annotation.NonNull;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Collection;
import com.boardgamegeek.util.MathUtils;
import com.boardgamegeek.util.StringUtils;

public class PlayCountFilterer extends CollectionFilterer {
	public static final int MIN_RANGE = 0;
	public static final int MAX_RANGE = 25;

	private int min;
	private int max;

	public PlayCountFilterer(Context context) {
		super(context);
	}

	public PlayCountFilterer(@NonNull Context context, int min, int max) {
		super(context);
		this.min = min;
		this.max = max;
	}

	@Override
	public void setData(@NonNull String data) {
		String[] d = data.split(DELIMITER);
		min = d.length > 0 ? MathUtils.constrain(StringUtils.parseInt(d[0], MIN_RANGE), MIN_RANGE, MAX_RANGE) : MIN_RANGE;
		max = d.length > 1 ? MathUtils.constrain(StringUtils.parseInt(d[1], MAX_RANGE), MIN_RANGE, MAX_RANGE) : MAX_RANGE;
	}

	@Override
	public int getTypeResourceId() {
		return R.string.collection_filter_type_play_count;
	}

	@NonNull
	@Override
	public String flatten() {
		return String.valueOf(min) + DELIMITER + String.valueOf(max);
	}

	public int getMax() {
		return max;
	}

	public int getMin() {
		return min;
	}

	@Override
	public String getDisplayText() {
		String text;
		if (max >= MAX_RANGE) {
			text = min + "+";
		} else if (min == max) {
			text = String.valueOf(max);
		} else {
			text = min + "-" + max;
		}
		return text + " " + context.getString(R.string.plays);
	}

	@Override
	public String getSelection() {
		String selection;
		if (max >= MAX_RANGE) {
			selection = Collection.NUM_PLAYS + ">=?";
		} else {
			selection = "(" + Collection.NUM_PLAYS + ">=? AND " + Collection.NUM_PLAYS + "<=?)";
		}
		return selection;
	}

	@Override
	public String[] getSelectionArgs() {
		if (max >= MAX_RANGE) {
			return new String[] { String.valueOf(min) };
		} else {
			return new String[] { String.valueOf(min), String.valueOf(max) };
		}
	}
}

File: app/src/main/java/com/boardgamegeek/filterer/PlayCountFilterer.kt
package com.boardgamegeek.filterer

import android.content.Context
import com.boardgamegeek.R
import com.boardgamegeek.model.CollectionItem
import com.boardgamegeek.extensions.IntervalDelegate
import java.util.*

class PlayCountFilterer(context: Context) : CollectionFilterer(context) {
    var min by IntervalDelegate(LOWER_BOUND, LOWER_BOUND, UPPER_BOUND)
    var max by IntervalDelegate(UPPER_BOUND, LOWER_BOUND, UPPER_BOUND)

    override val typeResourceId = R.string.collection_filter_type_play_count

    override fun inflate(data: String) {
        data.split(DELIMITER).run {
            min = getOrNull(0)?.toIntOrNull() ?: LOWER_BOUND
            max = getOrNull(1)?.toIntOrNull() ?: UPPER_BOUND
        }
    }

    override fun deflate() = "$min$DELIMITER$max"

    override val iconResourceId: Int
        get() = R.drawable.ic_baseline_event_available_24

    override fun chipText(): String {
        return describeRange()
    }

    override fun description(): String {
        return "${describeRange(" - ")} ${context.getString(R.string.plays)}"
    }

    fun describeRange(rangeDelimiter: String = "-") = when {
        max >= UPPER_BOUND -> String.format(Locale.getDefault(), "%,d+", min)
        min == max -> String.format(Locale.getDefault(), "%,d", max)
        min <= LOWER_BOUND -> String.format(Locale.getDefault(), "<%,d", max)
        else -> String.format(Locale.getDefault(), "%,d$rangeDelimiter%,d", min, max)
    }

    override fun filter(item: CollectionItem): Boolean {
        return when {
            max >= UPPER_BOUND -> item.numberOfPlays >= min
            else -> item.numberOfPlays in min..max
        }
    }

    companion object {
        const val LOWER_BOUND = 0
        const val UPPER_BOUND = 25
    }
}
